src/ratelimit_manager.cpp

```diff
@@ -0,0 +1,338 @@
+}
+
+bool RateLimitManager::edit_rule_by_id(const uint64_t id, const rate_limit_rule_t & new_rule) {
+    std::unique_lock<std::shared_mutex> lock(rate_limit_mutex);
+
+    // Check if a rule exists for the given ID
+    if(rule_store.count(id) > 0) {
+        auto rule = rule_store.at(id);
+
+        // Remove rule from rate limit rule pointer
+        for(const auto &entry : rule.values) {
+            rate_limit_entries.erase(rate_limit_rule_entry_t{rule.resource_type,entry});
+        }
+
+        // Update rule in rule store
+        rule_store.at(id) = new_rule;
+
+        // Add rule from rule store to rate limit rule pointer
+        for(const auto &entry : new_rule.values) {
+            rate_limit_entries.insert({rate_limit_rule_entry_t{new_rule.resource_type,entry}, &rule_store.at(id)});
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+const std::vector<rate_limit_rule_t> RateLimitManager::get_all_rules() {
+    std::shared_lock<std::shared_mutex> lock(rate_limit_mutex);
+
+    // Get all rules in a vector
+    std::vector<rate_limit_rule_t> rules;
+    for(const auto &rule : rule_store) {
+        rules.push_back(rule.second);
+    }
+
+    return rules;
+}
+
+const std::vector<rate_limit_status_t> RateLimitManager::get_banned_entries(const RateLimitedResourceType resource_type) {
+
+    std::shared_lock<std::shared_mutex> lock(rate_limit_mutex);
+
+    std::vector < rate_limit_status_t > banned_entries;
+
+    for (auto & element: throttled_entries) {
+        if (element.second.resource_type == resource_type) {
+            banned_entries.push_back(element.second);
+        }
+    }
+    
+
+    // Get permanent bans
+    for (auto & element: rule_store) {
+        if (element.second.action == RateLimitAction::BLOCK && element.second.resource_type == resource_type) {
+            for (auto & entry : element.second.values) {
+                banned_entries.push_back(rate_limit_status_t{true,0,0,entry, resource_type});
+            }
+        }
+    }
+
+    return banned_entries;
+    
+}
+
+void RateLimitManager::clear_all() {
+    std::unique_lock<std::shared_mutex> lock(rate_limit_mutex);
+    rate_limit_request_counts.clear();
+    rate_limit_entries.clear();
+    throttled_entries.clear();
+    rule_store.clear();
+    last_rule_id = 0;
+}
+
+void RateLimitManager::temp_ban_entry_wrapped(const rate_limit_rule_entry_t& entry, const int64_t number_of_days) {
+
+    // Check if entry is already banned
+    if (throttled_entries.count(entry) > 0 && throttled_entries.at(entry).is_banned) {
+        return;
+    }
+
+    auto now = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+    // Add entry to throttled_entries for the given number of days
+    throttled_entries.insert({entry, rate_limit_status_t{true, now, now + (number_of_days * 24 * 60 * 60), entry.value, entry.resource_type}});
+
+
+    if(rate_limit_request_counts.contains(entry)){
+        // Reset counters for the given entry
```

Comment:
Since entry is throttled, we can just delete the entry from `rate_limit_request_counts`?